File: rust/src/utilities/wait.rs
//! Wait/sleep utilities for browser automation.
//!
//! This module provides utilities for waiting and sleeping
//! with optional abort signal support.

use crate::core::engine::{EngineAdapter, EngineError};
use crate::core::navigation::is_navigation_error;
use std::time::Duration;
use tokio_util::sync::CancellationToken;

/// Result of a wait operation.
#[derive(Debug, Clone)]
pub struct WaitResult {
    /// Whether the wait completed normally.
    pub completed: bool,
    /// Whether the wait was aborted.
    pub aborted: bool,
}

impl WaitResult {
    /// Create a completed result.
    pub fn completed() -> Self {
        Self {
            completed: true,
            aborted: false,
        }
    }

    /// Create an aborted result.
    pub fn aborted() -> Self {
        Self {
            completed: false,
            aborted: true,
        }
    }
}

/// Wait for a specified duration.
///
/// # Arguments
///
/// * `duration` - How long to wait
/// * `reason` - Reason for waiting (for logging)
///
/// # Returns
///
/// The wait result
pub async fn wait(duration: Duration, _reason: Option<&str>) -> WaitResult {
    tokio::time::sleep(duration).await;
    WaitResult::completed()
}

/// Wait for a specified duration with abort support.
///
/// # Arguments
///
/// * `duration` - How long to wait
/// * `cancel_token` - Cancellation token to abort the wait
/// * `reason` - Reason for waiting (for logging)
///
/// # Returns
///
/// The wait result
pub async fn wait_with_cancel(
    duration: Duration,
    cancel_token: &CancellationToken,
    _reason: Option<&str>,
) -> WaitResult {
    // Check if already cancelled
    if cancel_token.is_cancelled() {
        return WaitResult::aborted();
    }

    tokio::select! {
        _ = tokio::time::sleep(duration) => {
            WaitResult::completed()
        }
        _ = cancel_token.cancelled() => {
            WaitResult::aborted()
        }
    }
}

/// Evaluate JavaScript in the page context.
///
/// # Arguments
///
/// * `adapter` - The engine adapter to use
/// * `script` - The JavaScript to evaluate
///
/// # Returns
///
/// The result of the evaluation
pub async fn evaluate(
    adapter: &dyn EngineAdapter,
    script: &str,
) -> Result<serde_json::Value, EngineError> {
    adapter.evaluate(script).await
}

/// Safe evaluate that catches navigation errors and returns a default value.
///
/// # Arguments
///
/// * `adapter` - The engine adapter to use
/// * `script` - The JavaScript to evaluate
/// * `default` - Default value to return on navigation error
///
/// # Returns
///
/// A result indicating success/failure and the value
pub async fn safe_evaluate(
    adapter: &dyn EngineAdapter,
    script: &str,
    default: serde_json::Value,
) -> SafeEvaluateResult {
    match adapter.evaluate(script).await {
        Ok(value) => SafeEvaluateResult {
            success: true,
            value,
            navigation_error: false,
        },
        Err(e) if is_navigation_error(&e.to_string()) => SafeEvaluateResult {
            success: false,
            value: default,
            navigation_error: true,
        },
        Err(_) => SafeEvaluateResult {
            success: false,
            value: default,
            navigation_error: false,
        },
    }
}

/// Result of a safe evaluate operation.
#[derive(Debug, Clone)]
pub struct SafeEvaluateResult {
    /// Whether the evaluation succeeded.
    pub success: bool,
    /// The value (actual result or default).
    pub value: serde_json::Value,
    /// Whether a navigation error occurred.
    pub navigation_error: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wait_result_completed() {
        let result = WaitResult::completed();
        assert!(result.completed);
        assert!(!result.aborted);
    }

    #[test]
    fn wait_result_aborted() {
        let result = WaitResult::aborted();
        assert!(!result.completed);
        assert!(result.aborted);
    }

    #[tokio::test]
    async fn wait_completes() {
        let start = std::time::Instant::now();
        let result = wait(Duration::from_millis(50), Some("test wait")).await;
        let elapsed = start.elapsed();

        assert!(result.completed);
        assert!(!result.aborted);
        assert!(elapsed >= Duration::from_millis(50));
    }

    #[tokio::test]
    async fn wait_with_cancel_completes_normally() {
        let token = CancellationToken::new();
        let result = wait_with_cancel(
            Duration::from_millis(10),
            &token,
            Some("test wait"),
        )
        .await;

        assert!(result.completed);
        assert!(!result.aborted);
    }

    #[tokio::test]
    async fn wait_with_cancel_aborts_on_cancel() {
        let token = CancellationToken::new();

        // Cancel immediately
        token.cancel();

        let result = wait_with_cancel(
            Duration::from_secs(10), // Long duration that would be cancelled
            &token,
            Some("test wait"),
        )
        .await;

        assert!(!result.completed);
        assert!(result.aborted);
    }

    #[tokio::test]
    async fn wait_with_cancel_aborts_during_wait() {
        let token = CancellationToken::new();
        let token_clone = token.clone();

        // Spawn task to cancel after a short delay
        tokio::spawn(async move {
            tokio::time::sleep(Duration::from_millis(50)).await;
            token_clone.cancel();
        });

        let start = std::time::Instant::now();
        let result = wait_with_cancel(
            Duration::from_secs(10), // Long duration that should be cancelled
            &token,
            Some("test wait"),
        )
        .await;
        let elapsed = start.elapsed();

        assert!(!result.completed);
        assert!(result.aborted);
        // Should have been cancelled quickly, not after 10 seconds
        assert!(elapsed < Duration::from_secs(1));
    }

    #[test]
    fn safe_evaluate_result_success() {
        let result = SafeEvaluateResult {
            success: true,
            value: serde_json::json!(42),
            navigation_error: false,
        };
        assert!(result.success);
        assert!(!result.navigation_error);
        assert_eq!(result.value, serde_json::json!(42));
    }

    #[test]
    fn safe_evaluate_result_navigation_error() {
        let result = SafeEvaluateResult {
            success: false,
            value: serde_json::Value::Null,
            navigation_error: true,
        };
        assert!(!result.success);
        assert!(result.navigation_error);
        assert!(result.value.is_null());
    }
}

File: js/src/utilities/wait.js
import { isNavigationError } from '../core/navigation-safety.js';
import { createEngineAdapter } from '../core/engine-adapter.js';

/**
 * Wait/sleep for a specified time with optional verbose logging
 * Now supports abort signals to interrupt the wait when navigation occurs
 *
 * @param {Object} options - Configuration options
 * @param {Function} options.log - Logger instance
 * @param {number} options.ms - Milliseconds to wait
 * @param {string} options.reason - Reason for waiting (for verbose logging)
 * @param {AbortSignal} options.abortSignal - Optional abort signal to interrupt wait
 * @returns {Promise<{completed: boolean, aborted: boolean}>}
 */
export async function wait(options = {}) {
  const { log, ms, reason, abortSignal } = options;

  if (!ms) {
    throw new Error('ms is required in options');
  }

  if (reason) {
    log.debug(() => `🔍 [VERBOSE] Waiting ${ms}ms: ${reason}`);
  }

  // If abort signal provided, use abortable wait
  if (abortSignal) {
    // Check if already aborted
    if (abortSignal.aborted) {
      log.debug(
        () => `🛑 Wait skipped (already aborted): ${reason || 'no reason'}`
      );
      return { completed: false, aborted: true };
    }

    return new Promise((resolve) => {
      let timeoutId = null;
      let abortHandler = null;

      const cleanup = () => {
        if (timeoutId) {
          clearTimeout(timeoutId);
        }
        if (abortHandler) {
          abortSignal.removeEventListener('abort', abortHandler);
        }
      };

      abortHandler = () => {
        cleanup();
        log.debug(() => `🛑 Wait aborted: ${reason || 'no reason'}`);
        resolve({ completed: false, aborted: true });
      };

      abortSignal.addEventListener('abort', abortHandler);

      timeoutId = setTimeout(() => {
        cleanup();
        if (reason) {
          log.debug(() => `🔍 [VERBOSE] Wait complete (${ms}ms)`);
        }
        resolve({ completed: true, aborted: false });
      }, ms);
    });
  }

  // Standard non-abortable wait (backwards compatible)
  await new Promise((r) => setTimeout(r, ms));

  if (reason) {
    log.debug(() => `🔍 [VERBOSE] Wait complete (${ms}ms)`);
  }

  return { completed: true, aborted: false };
}

/**
 * Evaluate JavaScript in page context
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Function} options.fn - Function to evaluate
 * @param {Array} options.args - Arguments to pass to function (default: [])
 * @param {Object} options.adapter - Engine adapter (optional, will be created if not provided)
 * @returns {Promise<any>} - Result of evaluation
 */
export async function evaluate(options = {}) {
  const { page, engine, fn, args = [], adapter: providedAdapter } = options;

  if (!fn) {
    throw new Error('fn is required in options');
  }

  const adapter = providedAdapter || createEngineAdapter(page, engine);
  return await adapter.evaluateOnPage(fn, args);
}

/**
 * Safe evaluate that catches navigation errors and returns default value
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Function} options.fn - Function to evaluate
 * @param {Array} options.args - Arguments to pass to function (default: [])
 * @param {any} options.defaultValue - Value to return on navigation error (default: null)
 * @param {string} options.operationName - Name for logging (default: 'evaluate')
 * @param {boolean} options.silent - Don't log warnings (default: false)
 * @returns {Promise<{success: boolean, value: any, navigationError: boolean}>}
 */
export async function safeEvaluate(options = {}) {
  const {
    page,
    engine,
    fn,
    args = [],
    defaultValue = null,
    operationName = 'evaluate',
    silent = false,
  } = options;

  try {
    const value = await evaluate({ page, engine, fn, args });
    return { success: true, value, navigationError: false };
  } catch (error) {
    if (isNavigationError(error)) {
      if (!silent) {
        console.log(
          `⚠️  Navigation detected during ${operationName}, recovering gracefully`
        );
      }
      return { success: false, value: defaultValue, navigationError: true };
    }
    throw error;
  }
}
